subprojects/core/src/main/java/org/gradle/api/artifacts/ComponentMetadataBuilder.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts;
+
+import org.gradle.api.Incubating;
+
+import java.util.List;
+
+/**
+ * A component metadata builder.
+ *
+ * @since 3.5
+ */
+@Incubating
+public interface ComponentMetadataBuilder {
```

Comment:
`ComponentMetadataDetails` contains more than what we need. The `Builder` is to build metadata. The details could actually extend the builder version. The idea is to separate the concerns: a builder that allows generating the minimal set of information we need.